File: backend/src/main/java/com/dominoes/lobby/websocket/OutgoingMessage.java
package com.dominoes.lobby.websocket;

import com.dominoes.lobby.dto.GameStateDto;
import com.dominoes.lobby.dto.LobbyStateDto;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public record OutgoingMessage(
        String type,
        Map<String, Object> payload
) {
    public static OutgoingMessage lobbyState(LobbyStateDto state) {
        return new OutgoingMessage("LOBBY_STATE", Map.of(
                "lobbyId", state.lobbyId() != null ? state.lobbyId().toString() : "",
                "size", state.size(),
                "users", state.users()
        ));
    }

    public static OutgoingMessage gameState(GameStateDto state) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("inProgress", state.inProgress());
        payload.put("boneyardCount", state.boneyardCount());
        payload.put("hand", state.hand());
        payload.put("currentPlayer", state.currentPlayerId() != null ? state.currentPlayerId().toString() : "");
        payload.put("openingPiece", state.openingPiece() != null ? state.openingPiece() : "");
        return new OutgoingMessage("GAME_STATE", payload);
    }

    public static OutgoingMessage joinAck(UUID userId) {
        return new OutgoingMessage("JOIN_ACK", Map.of("userId", userId.toString()));
    }

    public static OutgoingMessage startGameAck() {
        return new OutgoingMessage("START_GAME_ACK", Map.of());
    }

    public static OutgoingMessage endGameAck() {
        return new OutgoingMessage("END_GAME_ACK", Map.of());
    }

    public static OutgoingMessage error(String message) {
        return new OutgoingMessage("ERROR", Map.of("message", message));
    }
}

File: backend/internal/websocket/outgoing_message.go
package websocket

import (
	"github.com/dominoes/lobby/internal/dto"
	"github.com/google/uuid"
)

type OutgoingMessage struct {
	Type    string         `json:"type"`
	Payload map[string]any `json:"payload"`
}

func LobbyState(state dto.LobbyStateDto) OutgoingMessage {
	lobbyID := ""
	if state.LobbyID != nil {
		lobbyID = state.LobbyID.String()
	}
	return OutgoingMessage{
		Type: "LOBBY_STATE",
		Payload: map[string]any{
			"lobbyId": lobbyID,
			"size":    state.Size,
			"users":   state.Users,
		},
	}
}

func GameState(state dto.GameStateDto) OutgoingMessage {
	currentPlayer := ""
	if state.CurrentPlayerID != nil {
		currentPlayer = state.CurrentPlayerID.String()
	}
	table := make([]map[string]string, len(state.Table))
	for i, piece := range state.Table {
		table[i] = map[string]string{
			"code":     piece.Code,
			"rotation": piece.Rotation,
		}
	}
	return OutgoingMessage{
		Type: "GAME_STATE",
		Payload: map[string]any{
			"inProgress":      state.InProgress,
			"boneyardCount":   state.BoneyardCount,
			"hand":            state.Hand,
			"currentPlayer":   currentPlayer,
			"table":           table,
			"drawnThisTurn":   state.DrawnThisTurn,
		},
	}
}

func JoinAck(userID uuid.UUID) OutgoingMessage {
	return OutgoingMessage{
		Type:    "JOIN_ACK",
		Payload: map[string]any{"userId": userID.String()},
	}
}

func StartGameAck() OutgoingMessage {
	return OutgoingMessage{Type: "START_GAME_ACK", Payload: map[string]any{}}
}

func PlayPieceAck() OutgoingMessage {
	return OutgoingMessage{Type: "PLAY_PIECE_ACK", Payload: map[string]any{}}
}

func DrawFromBoneyardAck() OutgoingMessage {
	return OutgoingMessage{Type: "DRAW_FROM_BONEYARD_ACK", Payload: map[string]any{}}
}

func EndGameAck() OutgoingMessage {
	return OutgoingMessage{Type: "END_GAME_ACK", Payload: map[string]any{}}
}

func ErrorMessage(message string) OutgoingMessage {
	return OutgoingMessage{
		Type:    "ERROR",
		Payload: map[string]any{"message": message},
	}
}
